scp: do command STDERR pipe-ing for large output

mxmCherry opened the pull request:
Addresses https://github.com/bsm/bfs.rb/issues/28

cc @dim @vrih

mxmCherry committed:
scp: do command STDERR pipe-ing for large output

mxmCherry commented on docker-compose.yml:
```diff
@@ -0,0 +1,15 @@
+  ssh:
+    image: sickp/alpine-sshd:7.5-r2
+    ports:
+      - '7022:22'
```
`docker-compose.yml` is just for localhost development/debugging

mxmCherry commented on scp/lib/bfs/bucket/scp.rb:
```diff
@@ -145,32 +145,53 @@ def mkdir_p(path)
         cmdstr = cmd.map {|x| Shellwords.escape(x) }.join(' ')
 
+        # init command's STDOUT stream (large output may break lines in on_data):
+        stdout_reader, stdout_writer = IO.pipe
```
I couldn't find anything more proper (in-memory stream piping like Go's https://golang.org/pkg/io/#Pipe) - Ruby's IO.pipe looks to be backed by OS pipes:

https://ruby-doc.org/core-3.0.0/IO.html#method-c-pipe 

> Not available on all platforms.

Heavy or not, it should do the trick. Maybe we could swap it with some custom-written in-mem pipe (haven't found anything like that for Ruby) if that ever becomes a concern.

dim replied:
I don't think you need to do it, you can do something much simpler and accept a block to allow chunking:

```ruby
def sh!(*cmd)
...
   ch.on_data do |_, data|
      yield data
   end
```

basically, instead of concatting into stdout and returning it, let the caller process chunks of stdout instead

mxmCherry replied:
@dim that was the original approach, and lines were broken for large output. Hence streaming (stream can combine 2 chunks with broken line in the middle, so each_line works as expected)

dim replied:
k, checked again, I still don't like the abuse of pipe that is happening here. it doesn't need to be that complicated, you can just do something like this:

```rb
ch.on_data do |_, data|
  stdout << data
  if block_given?
    pos = stdout.index("\n")    
    yield stdout.slice!(..pos) if pos
  end
end
```

mxmCherry replied:
:man_facepalming: right

mxmCherry commented on scp/lib/bfs/bucket/scp.rb:
```diff
@@ -145,32 +145,53 @@ def mkdir_p(path)
         stdout
+      ensure
+        stdout_writer&.close
+        stdout_reader&.close
```
You can work with pipes using block (which ensures that stuff is all closed):

```ruby
IO.pipe do |reader, writer|
  ...
end
```

But I thought it'll be too deep nesting, so decided to follow `reader, writer = IO.pipe` approach.

mxmCherry commented on scp/lib/bfs/bucket/scp.rb:
```diff
@@ -145,32 +145,53 @@ def mkdir_p(path)
-        @client.session.loop
+
+        # loop in background (blocking to be done by consumer):
+        loop_thread = Thread.new { @client.session.loop }
```
Had to move that in thread 'cause `.loop` is blocking, needed to fit 2 blocking processes (loop + caller read) together.

dim commented on scp/lib/bfs/bucket/scp.rb:
```diff
@@ -145,32 +145,53 @@ def mkdir_p(path)
-              if block_given?
-                yield data
-              else
-                stdout += data
```
that was my mistake, we should not reallocate but append:  `<<`

mxmCherry committed:
scp: simplify large stdout each_line handling

mxmCherry commented on Gemfile:
```diff
@@ -9,3 +9,4 @@ gem 'rake'
 gem 'rspec'
 gem 'rubocop'
 gem 'rubocop-performance'
+gem 'webrick'
```
That's for Ruby 3: https://www.ruby-lang.org/en/news/2020/12/25/ruby-3-0-0-released/ (removed from stdlib, need to install separately - looks to be implicitly used by s3 or so)

vrih commented on scp/lib/bfs/bucket/scp.rb:
```diff
@@ -153,10 +153,11 @@ def sh!(*cmd) # rubocop:disable Metrics/MethodLength
-              else
-                stdout += data
+                pos = stdout.rindex("\n")
+                yield stdout.slice!(..pos) if pos
```
Failing rubocop

mxmCherry committed:
hk: rubocop fixes

mxmCherry committed:
hk: switch to rubocop-bsm

mxmCherry committed:
hk: rubocop - disable RSpec/MessageSpies for now

RSpec/MessageSpies: Prefer have_received for setting message expectations. Setup bucket as a spy using allow or instance_spy.
      expect(bucket).to receive(:close)

(and 3 more of these)

mxmCherry commented on .rubocop.yml:
```diff
@@ -1,9 +1,19 @@
+    - core/spec/bfs/helpers_spec.rb
+
+RSpec/MessageSpies:
+  Enabled: false
```
```
core/spec/bfs_spec.rb:13:25: C: RSpec/MessageSpies: Prefer have_received for setting message expectations. Setup bucket as a spy using allow or instance_spy.
      expect(bucket).to receive(:close)
```

I think suggested usage is a bit too verbose, but can change to what it suggests (allow and then expect... have_received)

dim commented on scp/lib/bfs/bucket/scp.rb:
```diff
@@ -167,6 +168,12 @@ def sh!(*cmd) # rubocop:disable Metrics/MethodLength
+
+        if block_given? && stdout.length.positive?
+          yield stdout
+          stdout = ''
```
or free of allocations: `stdout.clear`

mxmCherry committed:
scp: minor - string.clear instead of string = ''

mxmCherry commented:
@dim to merge/release
